posts/Stef/2019-11-18-hibernate-orm-with-panache-in-quarkus.adoc

```diff
@@ -0,0 +1,462 @@
+    }
+
+    @GET
+    @Path("people/born-after")
+    public List<Person> findBornAfter(@PathParam Date date){
+        return personDao.findBornAfter(date);
+    }
+
+    @GET
+    @Path("person/{id}")
+    public Person findById(@PathParam Long id){
+        Person p = personDao.findById(id);
+        if(p == null)
+            throw new WebApplicationException(Status.NOT_FOUND);
+        return p;
+    }
+
+    @PUT
+    @Path("person/{id}")
+    public void updatePerson(@PathParam Long id, Person newPerson){
+        Person p = personDao.findById(id);
+        if(p == null)
+            throw new WebApplicationException(Status.NOT_FOUND);
+        p.setBirth(newPerson.getBirth());
+        p.setFirstName(newPerson.getFirstName());
+        p.setLastName(newPerson.getLastName());
+    }
+
+    @DELETE
+    @Path("person/{id}")
+    public void deletePerson(@PathParam Long id){
+        Person p = personDao.findById(id);
+        if(p == null)
+            throw new WebApplicationException(Status.NOT_FOUND);
+        personDao.delete(p);
+    }
+    
+    @POST
+    @Path("people")
+    public Response newPerson(@Context UriInfo uriInfo, Person newPerson){
+        Person p = new Person();
+        p.setBirth(newPerson.getBirth());
+        p.setFirstName(newPerson.getFirstName());
+        p.setLastName(newPerson.getLastName());
+        personDao.persist(p);
+        
+        URI uri = uriInfo.getAbsolutePathBuilder()
+                .path(PersonEndpoint.class)
+                .path(PersonEndpoint.class, "findById")
+                .build(p.getId());
+        return Response.created(uri).build();
+    }
+}
+----
+
+
+== A few observations on the traditionnal JPA way
+
+We've all seen hundreds of entities and DAOs written this way. They don't have anything surprising.
+
+What they do have plenty of, on the other hand is boilerplate:
+
+- The generated ID field. Very often, all your entities will use the same auto-generate ID type.
+- All those property accessors that do nothing in the entity. They are required for encapsulation, and because Java
+does not support first-class properties in the language. Most people either generate them from their IDE, or using
+https://projectlombok.org/features/GetterSetter[Lombok].
+- All those `persist`/`delete`/`findById`/`findAll` methods on every DAO. All DAOs have them.
+- Those DAO queries all start with `FROM Person` and have to repeat the `Person.class` all over the place.
+
+== A first look at what Hibernate ORM with Panache can do for us
+
+Let's jump forward to https://quarkus.io/[Quarkus] and in particular, what 
+https://quarkus.io/guides/hibernate-orm-panache[Hibernate ORM with Panache] can do for us. It turns out to be quite a lot.
+
+Quarkus allows us to do a lot of bytecode modification at build-time, which (among many benefits) allows us to side-step
+Java's lack of support for first-class properties by:
+
+- Writing public fields instead of private+getter+setter
```

Comment:
Btw. does it also do the bi-di association management the traditional Hibernate enhancement plug-in could do?

Reply:
Yes, it doesn't add or remove any Hibernate feature. What makes you think it would remove that?

Reply:
The way I've sawn it being used was via a build time Maven plug-in, wasn't sure whether that'd be usable in case of Quarkus. Cool if it is. Thanks for clarifying!

Reply:
It's using the normal hibernate bytecode enhancer like Hibernate ORM on Quarkus, which is not as tested as the runtime enhancements, but still nothing related to Panache itself.

Reply:
@gunnarmorling yes that's enabled by default (and can't be disabled!)